text/0000-safety-tags.md

```diff
@@ -0,0 +1,662 @@
+- Feature Name: safety_tag
+- Start Date: 2025-07-29
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC introduces a concise safety-comment convention for unsafe code in standard libraries:
+tag every public unsafe function with `#[safety::requires]` and call with `#[safety::checked]`.
+
+Safety tags refine today’s safety-comment habits: a featherweight syntax that condenses every
+requirement into a single, check-off reminder.
+
+The following snippet [compiles] today if we enable enough nightly features, but we expect Clippy
+and Rust-Analyzer to enforce tag checks and provide first-class IDE support.
+
+[compiles]: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=6eb0e47c416953da1f2470b11417e69a
+
+```rust
+#[safety::requires { // 💡 define safety tags on an unsafe function
+    ValidPtr = "src must be [valid](https://doc.rust-lang.org/std/ptr/index.html#safety) for reads",
+    Aligned = "src must be properly aligned, even if T has size 0",
+    Initialized = "src must point to a properly initialized value of type T"
+}]
+pub unsafe fn read<T>(ptr: *const T) { }
+
+fn main() {
+    #[safety::checked { // 💡 discharge safety tags on an unsafe call
+        ValidPtr, Aligned, Initialized = "optional reason"
+    }]
+    unsafe { read(&()) };
+}
+```
+
+# Motivation
+[motivation]: #motivation
+
+## Safety Invariants: Forgotten by Authors, Hidden from Reviewers
+
+To avoid the misuse of unsafe code, Rust developers are encouraged to provide clear safety comments
+for unsafe APIs. Safety comments are often repetitive and may be perceived as less important
+than the code itself, which makes them error-prone and increases the risk that reviewers may
+overlook inaccuracies or missing safety requirements.
+
+Recent Rust issues [#134496],[#135805], and [#135009] illustrate the problem: several libstd APIs
+silently rely on the global allocator, yet their safety comments never state the crucial invariant
+that any raw pointer passed to them must refer to memory allocated by that same global allocator.
+The pattern is always the same.
+
+```rust
+// Ptr is possibly from another allocation.
+pub unsafe fn from_raw(ptr: *const T) -> Self {
+    // SmartPoiner can be Box, Arc, Rc, and Weak.
+    unsafe { SmartPoiner::from_raw_in(ptr, Global) }
+}
+```
+
+[#134496]: https://github.com/rust-lang/rust/pull/134496
+[#135805]: https://github.com/rust-lang/rust/pull/135805
+[#135009]: https://github.com/rust-lang/rust/pull/135009
+
+Even if the safety documentation is complete, two problems remain:
```

Comment:
There is another important benefit that should be called out here: what happens if the safety invariants change? Updating callsites (including in downstream consumer code) is very important and extremely easy to overlook.

Reply:
Yeah. That's what I wrote in [Safety Invariants Have No Semver](https://github.com/Artisan-Lab/rfcs/blob/safety-tags/text/0000-safety-tags.md#safety-invariants-have-no-semver).